File: SwingApps/src/cc/lib/swing/AWTNumberPicker.java
package cc.lib.swing;

import javax.swing.JLabel;
import javax.swing.JSpinner;
import javax.swing.SpinnerNumberModel;
import javax.swing.event.ChangeEvent;
import javax.swing.event.ChangeListener;

import cc.lib.game.Utils;


public class AWTNumberPicker extends AWTPanel implements ChangeListener {

    public interface Listener {
        void onValueChanged(int oldValue, int newValue);
    };

    private SpinnerNumberModel model;
    private Listener listener;

    public AWTNumberPicker(int rows, int cols) {
        super(rows, cols);
    }

    public AWTNumberPicker() {
        super();
    }

    public int getValue() {
        return (Integer)model.getValue();
    }

    public void setValue(int value) {
        ignore = true;
        model.setValue(value);
        ignore = false;
    }

    private boolean ignore = false;

    public final void stateChanged(ChangeEvent e) {
        if (!ignore && listener != null)
            listener.onValueChanged(0, (Integer)model.getValue());
    }

    public static class Builder {

        private int min=Integer.MIN_VALUE;
        private int max=Integer.MAX_VALUE;
        private int value = 0;
        private int step = 1;
        private String label = null;

        public AWTNumberPicker build(Listener listener) {
            AWTNumberPicker panel;
            if (label != null) {
                panel = new AWTNumberPicker(0, 1);
                panel.add(new JLabel(label));
            } else {
                panel = new AWTNumberPicker();
            }
            value = Utils.clamp(value, min, max);
            panel.model = new SpinnerNumberModel(value, min, max, step);
            JSpinner spinner = new JSpinner(panel.model);
            spinner.addChangeListener(panel);
            panel.add(spinner);
            panel.listener = listener;
            return panel;
        }

        public Builder setMin(int min) {
            this.min = min;
            return this;
        }

        public Builder setMax(int max) {
            this.max = max;
            return this;
        }

        public Builder setValue(int value) {
            this.value = value;
            return this;
        }

        public Builder setStep(int step) {
            this.step = step;
            return this;
        }

        public Builder setLabel(String label) {
            this.label = label;
            return this;
        }
    }

}

File: SwingApps/src/cc/lib/swing/AWTNumberPicker.kt
package cc.lib.swing

import javax.swing.JLabel
import javax.swing.JSpinner
import javax.swing.SpinnerNumberModel
import javax.swing.event.ChangeEvent
import javax.swing.event.ChangeListener

class AWTNumberPicker internal constructor(
	private val model: SpinnerNumberModel,
	private val callback: ((Int) -> Unit)?,
	rows: Int, cols: Int
) : AWTPanel(rows, cols), ChangeListener {

	var value: Int
		get() = model.value as Int
		set(value) {
			ignore = true
			model.value = value
			ignore = false
		}
	private var ignore = false
	override fun stateChanged(e: ChangeEvent) {
		if (!ignore) callback?.invoke(model.value as Int)
	}

	class Builder {
		private var min = Int.MIN_VALUE
		private var max = Int.MAX_VALUE
		private var value = 0
		private var step = 1
		private var label: String? = null
		fun build(callback: ((Int) -> Unit)?): AWTNumberPicker {
			val model = SpinnerNumberModel(value, min, max, step)
			val rows = 0
			val cols = if (label != null) 1 else 0
			return AWTNumberPicker(model, callback, rows, cols).also { panel ->
				label?.let {
					panel.add(JLabel(it))
					val spinner = JSpinner(panel.model)
					spinner.addChangeListener(panel)
					panel.add(spinner)
				}
			}
		}

		fun setMin(min: Int): Builder {
			this.min = min
			return this
		}

		fun setMax(max: Int): Builder {
			this.max = max
			return this
		}

		fun setValue(value: Int): Builder {
			this.value = value
			return this
		}

		fun setStep(step: Int): Builder {
			this.step = step
			return this
		}

		fun setLabel(label: String?): Builder {
			this.label = label
			return this
		}
	}
}